### PR TITLE
Added Mouse Events

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

Related with this issue https://github.com/ignitionrobotics/ign-gui/issues/209. I added left and right mouse events. These events are generated by the Scene3D window, otherwise if we try to get the mouse click from other plugin we will get the click in the hole window making more difficult to handle it. Other advantage it's that we are publishing a ignition related event.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

**Review by mjcarroll — APPROVED** [review]
LGTM, but the `INTEGRATION_Scene3d` test is segfaulting now.

**Comment by mjcarroll** [comment]
@osrf-jenkins retest this please

File: include/ignition/gui/GuiEvents.hh [hunk 1]
```diff
@@ -255,6 +259,59 @@ namespace ignition
...
+        /// \param[in] _mouse The left mouse event
+        /// the scene
+        public: explicit LeftClickOnScene(
+          const ignition::common::MouseEvent &_mouse)
+            : QEvent(kType), mouse(_mouse)
+        {
+        }
+
+        /// \brief Unique type for this event.
+        static const QEvent::Type kType = QEvent::Type(QEvent::MaxUser - 10);
```
Comment (chapulina) [inline include/ignition/gui/GuiEvents.hh:277]: Why skip `-8` and `-9`?

**Review by chapulina — COMMENTED** [review]
Have you looked into the `INTEGRATION_Scene3d` test failure?

File: include/ignition/gui/GuiEvents.hh (hunk 1 above)
Comment (chapulina) [reply include/ignition/gui/GuiEvents.hh:277]: Oh I see them in https://github.com/ignitionrobotics/ign-gui/pull/224

**Review by chapulina — CHANGES_REQUESTED** [review]
Do you have a PR that uses these events, @ahcorde ? It would be nice to make sure the events have everything we need before merging and releasing. These classes don't use the PIMPL pattern, so if we realize later that they need more fields, we can't break ABI. See https://github.com/ignitionrobotics/ign-gui/issues/209#issuecomment-849808255 for more context.

**Comment by ahcorde** [comment]
@chapulina this PR make use of it https://github.com/ignitionrobotics/ign-gui/pull/231/

**Comment by ahcorde** [comment]
friendly ping @chapulina

**Comment by chapulina** [comment]
@ahcorde , how about we target this one at `main` so we can iterate on it a bit. Then if we see the need to backport we can. It doesn't look like we have a use case for this on Citadel yet, right?

**Comment by ahcorde** [comment]
I will wait to this forward port https://github.com/ignitionrobotics/ign-gui/pull/238 because it includes the scene3d tests.

**Comment by ahcorde** [comment]
@chapulina retargeted to `main` and add PIMPL pattern

**Review by chapulina — APPROVED** [review]
I pushed some documentation and tweaks in edcfb77be72a9f25b7edd4f5043c4e1d7d873fc4.

I'm on the fence about keeping both the `ClickOnScene` and `ClickToScene` events. I think they could be combined into a single event to avoid confusion. But I'll refrain from making more changes until @scpeters opens a PR adding the distance field as proposed in https://github.com/ignitionrobotics/ign-gui/issues/209#issuecomment-849808255.

Let's go with these for now!

**Pull request merged**